### PR TITLE
[Manager] Added limits to validating doubles in preferences

### LINKED ISSUE #4650: Large number in "switch between applications" is misunderstood
If you enter 1000000 minutes into "switch between applications", it saves ok, but next time this is displayed as 1+e006, and saving it like that tells me it's an invalid number.

### PR DESCRIPTION
**Pull request by Vulpine05**
Fixes #4650 

**Description of the Change**
Changed the function of validating some computing preferences from "IsValidFloat" to "IsValidFloatBetween".  Added upper limit of 10,000 to prevent display errors when reading the stored preferences.

**Alternate Designs**
Refer to the discussion in #4650.

**Release Notes**
Upper limits created for some computing preferences.

### CODE REVIEW

**Comment by AenBleidd** [comment]
I'm afraid hard limit of 10k is not nice. Sometimes it's really necessary to set more than 10k.
I'll keep it open for a while for a further discussion.

**Comment by Vulpine05** [comment]
> I'm afraid hard limit of 10k is not nice. Sometimes it's really necessary to set more than 10k. I'll keep it open for a while for a further discussion.

Fair enough, but I was matching the upper limits that were already set for other variables.  For example, maximum disk space.  I  don't think there is a problem setting the upper limit higher, as long as it doesn't overflow the variable when it gets stored (which is around 21,474,836.00).  If we change this upper limit, it should be for the ones in the current commit and all other variables checked against the 10k limit.  I suppose there could be possibilities that we would want to have some variables have different upper limits, but off the top of my head I don't know what that would be.  Setting the upper limit higher could possibly cause issues.  I can chase those variables down to see if there are any limits.

**Comment by AenBleidd** [comment]
@Vulpine05, I just can't say for sure which limits should be set.
E.g., 10k seconds is not enough is definitely not enough to set for checkpoints interval if user wants to disable it, because it's less than 3 hours.
So every parameter should have their own limit, and it should be discussed, what limit should be set in every case.

**Comment by Vulpine05** [comment]
> So every parameter should have their own limit, and it should be discussed, what limit should be set in every case.
I originally agreed with this statement, but then I remembered that this is a validation check on the input.  If this is checking only to see if the input is a valid double, then should we really ask it to do more?  I would assume after the settings are validated, then the specific settings are evaluated based on the limits of the settings, if needed?  I don't know, I will have to walk the code.  I can look into this and see what happens.

On a somewhat related note, we probably need to see how these settings are validated with web preferences.

**Comment by CharlieFenton** [comment]
The preferences dialog code contains two methods for validating doubles. For simple preferences, they are:
```
bool CPanelPreferences::IsValidFloatValue(const wxString& value, bool allowNegative)
bool CPanelPreferences::IsValidFloatValueBetween(const wxString& value, double minVal, double maxVal)

```
and for advanced preferences:
```
bool CDlgAdvPreferences::IsValidFloatValue(const wxString& value, bool allowNegative)
bool CDlgAdvPreferences::IsValidFloatValueBetween(const wxString& value, double minVal, double maxVal)
```
Once you have determined the desirable limits for each case, you can probably replace the `IsValidFloatValue` call with `IsValidFloatValueBetween`, passing in the appropriate limits for each case.

You will probably also want to modify the way `CDlgAdvPreferences::ValidateInput()` and `CPanelPreferences::ValidateInput()` display errors when a value is out of bounds to allow displaying different limits.

**Comment by Vulpine05** [comment]
@davidpanderson , @AenBleidd , this PR has been revised and is ready to review.  Some significant changes:

1. RoundToNearestHundredth function now uses __int64 instead of int.
2. I found that when saving values with 14 or more digits caused precision errors.  Therefore, I think our highest limit should be 13 digits for the integer.
3. wxWidgets has a validator to limit how many characters are in a wxTextCtrl.  I have included this in the validation to prevent a user from just holding down a key for a minute.  This will give the user some initial feedback (if they pay attention) that there is a limit to their input.  The limit is 16 characters (13 for the integer, 1 for the decimal point, 2 for the decimal places).
4. However, a user can still enter 14 or more digits, so each double variable still needs to have its value checked.  I have revised the upper limit from my previous commit to be higher.
5. Same changes above have been made to the two variables affected for the preferences in simple view.
6. Same changes made to daily_xfer_period_days, but since this is an int, the limits are 10 characters (instead of 16) and the maximum valid number is the maximum for an integer.

I have tested this in Windows 10 and any number modified in the Manager is saved and restored ALMOST the same.  There are some trailing digits that are saved in global_prefs_override, but when it is parsed back into the Manager that data is truncated.  If you feel the upper limit is too high, I would be happy to adjust, just let me know what you think that should be.

**Review by AenBleidd — CHANGES_REQUESTED** [review]

File: clientgui/DlgAdvPreferences.cpp [hunk 1]
```diff
@@ -235,8 +252,9 @@ wxString CDlgAdvPreferences::DoubleToTimeString(double dt) {
 // precision of saved values to .01.  This prevents unexpected
 // behavior when, for example, a zero value means no restriction
 // and the value is displayed as 0.00 but is actually 0.001.
+//
 double CDlgAdvPreferences::RoundToHundredths(double td) {
-    int i = (int)((td + .005) * 100.);
+    __int64 i = (__int64)((td + .005) * 100.);
```
[the remark this answers is not in the document]
Comment (Vulpine05) [reply clientgui/DlgAdvPreferences.cpp]: Fixed.

File: clientgui/sg_DlgPreferences.cpp [hunk 2]
```diff
@@ -614,8 +616,9 @@ wxString CPanelPreferences::DoubleToTimeString(double dt) {
 // precision of saved values to .01.  This prevents unexpected
 // behavior when, for example, a zero value means no restriction
 // and the value is displayed as 0.00 but is actually 0.001.
+//
 double CPanelPreferences::RoundToHundredths(double td) {
-    int i = (int)((td + .005) * 100.);
+    __int64 i = (__int64)((td + .005) * 100.);
```
[the remark this answers is not in the document]
Comment (Vulpine05) [reply clientgui/sg_DlgPreferences.cpp]: Fixed.

**Review by AenBleidd — APPROVED** [review]

**Comment by AenBleidd** [comment]
@Vulpine05, lgtm, thanks for the fix!

**Comment by Vulpine05** [comment]
You're welcome, @AenBleidd !  I see this PR was moved to 7.22.0.  The release notes for 7.20.0 pre-release still say the large number bug was fixed.  Sorry to nit-pick, but I think that needs to be edited.

**Comment by RichardHaselgrove** [comment]
I'm trying to talk a user through a similar discrepancy at the moment. The [Release Notes](https://boinc.berkeley.edu/wiki/Release_Notes) for v7.16.11 (September 2020) say:

- Client: let a project master URL change from http: to https: without involving the user.

But the code for PR #4539 wasn't written and merged until October 2021, over a year later.  I'm about to do a deep dive into the v7.16 release tags, to try and work out what exact code was used.

**Comment by RichardHaselgrove** [comment]
Cancel that. Looks like it's a case of "so good, we fixed it twice". There's an earlier PR #3753, which **did** appear in v7.16.11

**Pull request merged**